File: 02/sol.py
def read_file(filename="input.txt"):
    with open("input.txt", "r") as f:
        return f.read().strip()


def parse(input_string):
    lines = input_string.split("\n")
    chars = [tuple(line.split()) for line in lines]
    return chars


def gen_point_dict1():
    l1 = ["A", "B", "C"]
    l2 = ["X", "Y", "Z"]

    def func(c1, c2):
        i1 = l1.index(c1)
        i2 = l2.index(c2)
        return (i2+1) + (((((i2-i1)%3)+1)*3)%9)

    rtn = {}
    for c1 in l1:
        for c2 in l2:
            rtn[(c1, c2)] = func(c1, c2)
    return rtn


def solve1(chars, points=gen_point_dict1()):
    return sum(points[line] for line in chars)


def gen_point_dict2():
    l1 = ["A", "B", "C"]
    l2 = ["X", "Y", "Z"]

    def func(c1, c2):
        i1 = l1.index(c1)
        i2 = l2.index(c2)
        return (i2*3) + ((((i2-1)+i1)%3)+1)

    rtn = {}
    for c1 in l1:
        for c2 in l2:
            rtn[(c1, c2)] = func(c1, c2)
    return rtn


def solve2(chars, points=gen_point_dict2()):
    return sum(points[line] for line in chars)


if __name__ == "__main__":
    input_string = read_file("input.txt")
    assert input_string is not None
    chars = parse(input_string)
    res1 = solve1(chars)
    print(res1)
    res2 = solve2(chars)
    print(res2)

File: 02/sol.jl
function main()
    filename = isempty(ARGS) ? "input.txt" : ARGS[1]
    input = rstrip(read(filename, String))
    x = parseinput(input)
    r1 = solve1(x)
    println(r1)
    r2 = solve2(x)
    println(r2)
end

function parseinput(input)
    lines = split(input, "\n")
    chars = map(line -> (line[1], line[3]), lines)
    return chars
end

function func1(x, y)
    i = x - 'A'
    j = y - 'X'
    return (j+1) + mod((3*(1+mod(j-i, 3))), 9)
end

function gendict(func)
    # return Dict(((x,y), func(x, y)) for x in 'A':'C' for y in 'X':'Z')
    return Dict((t, func(t...)) for t in Iterators.product('A':'C', 'X':'Z'))
end

function solve1(x)
    d = gendict(func1)
    return sum(d[t] for t in x)
end

function func2(x, y)
    i = x - 'A'
    j = y - 'X'
    return (j*3) + (1+mod(j-1+i, 3))
end

function solve2(x)
    d = gendict(func2)
    return sum(d[t] for t in x)
end

main()
